dotnet/src/SemanticKernel.Abstractions/Contents/BinaryContent.cs

```diff
@@ -3,118 +3,310 @@
+        string? dataUri)
     {
-        Verify.NotNull(streamProvider, nameof(streamProvider));
+        this.DataUri = dataUri;
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="BinaryContent"/> class from a byte array.
+    /// </summary>
+    /// <param name="data">Byte array content</param>
+    /// <param name="mimeType">The mime type of the content</param>
+    public BinaryContent(
+        ReadOnlyMemory<byte> data,
+        string? mimeType)
+    {
+        Verify.NotNull(data, nameof(data));
+        if (data.IsEmpty)
+        {
+            throw new ArgumentException("Data cannot be empty", nameof(data));
+        }
 
-        this._streamProvider = streamProvider;
+        this.MimeType = mimeType;
+        this.Data = data;
     }
 
+    #region Private
     /// <summary>
-    /// Access the content stream.
+    /// Sets the Uri of the content.
     /// </summary>
-    /// <remarks>
-    /// Caller responsible for disposal.
-    /// </remarks>
-    public async Task<Stream> GetStreamAsync()
+    /// <param name="uri">Target Uri</param>
+    private void SetUri(Uri? uri)
     {
-        if (this._streamProvider is not null)
+        if (uri?.Scheme == "data")
         {
-            return await this._streamProvider.Invoke().ConfigureAwait(false);
+            throw new InvalidOperationException("For DataUri contents, use DataUri property.");
         }
 
-        if (this.Content is not null)
+        this._referencedUri = uri;
+    }
+
+    /// <summary>
+    /// Gets the Uri of the content.
+    /// </summary>
+    /// <returns>Uri of the content</returns>
+    private Uri? GetUri()
+        => this._referencedUri;
+
+    /// <summary>
+    /// Sets the DataUri of the content.
+    /// </summary>
+    /// <param name="dataUri">DataUri of the content</param>
+    /// <exception cref="ArgumentException"></exception>
+    private void SetDataUri(string? dataUri)
+    {
+        if (dataUri is null)
         {
-            return new MemoryStream(this.Content.Value.ToArray());
+            this._dataUri = null;
+
+            // Invalidate the current bytearray
+            this._data = null;
+            return;
         }
 
-        throw new KernelException("Null content");
+        var isDataUri = dataUri!.StartsWith("data:", StringComparison.OrdinalIgnoreCase) == true;
+        if (!isDataUri)
+        {
+            throw new UriFormatException("Invalid data uri. Scheme should start with \"data:\"");
+        }
+
+        // Validate the dataUri format
+        var parsedDataUri = DataUriParser.Parse(dataUri);
+
+        // Overwrite the mimetype to the DataUri.
+        this.MimeType = parsedDataUri.MimeType;
+
+        // If parameters where provided in the data uri, updates the content metadata.
+        if (parsedDataUri.Parameters.Count != 0)
+        {
+            this.UpdateDataUriParametersToMetadata(parsedDataUri);
```

Comment:
After researching for the `Data Uri` specification and understanding this is accepted, since we have a parser already implemented adding this made sense when using the `Data Uri` on its full glance and seems handy way of not losing data or enforcing the caller to interate over those to add manually to metadata. 
The parser also has a special logic to only interate over data uri parameters if any is present.